pkg/xds/server/components.go

```diff
@@ -271,6 +271,20 @@ func DefaultDataplaneSyncTracker(rt core_runtime.Runtime, reconciler, ingressRec
 					return err
 				}
 
+				// Need to get dataplane again after counting hash. Because the following situation is possible:
+				// 1. Get Dataplane(version 1)
+				// 2. Dataplane is externally updated to version 2
+				// 3. Calculate hash in meshSnapshotCache.GetHash. This function list all resources and hash will be built based on Dataplane(version 2)
+				// 4. Use Dataplane(version 1) when generate Envoy resources
+				//
+				// So update from version 1 to version 2 is effectively lost.
+				if err := rt.ReadOnlyResourceManager().Get(ctx, dataplane, core_store.GetBy(key)); err != nil {
+					if core_store.IsResourceNotFound(err) {
+						return reconciler.Clear(&proxyID)
+					}
+					return err
+				}
```

Comment:
I don't think this fix the situation you described, because you used `ReadOnlyResourceManager`.
When you get the dataplane in the first step, it is cached so it's very high chance you will still retrieve V1. I think you need to either
1. Use non cached ResourceManager
2. Use the dataplane that was used to compiled the hash

Reply:
This should be created as a separate PR imho

Reply:
If you get Dataplane after calculating hash it is the same or newer than a dataplane which reflected in the hash. Because when calculating hash we are also using ReadOnlyResourceManager

Reply:
@jakubdyszkiewicz create separate PR https://github.com/kumahq/kuma/pull/1313, let's proceed there